File: compare-sort-algorithms/src/radix_sort.rs
use std::fs::File;
use std::io::{self, BufRead, BufReader, Write};
use std::time::Instant;
use std::env;

/// Radix Sort implementation in Rust.
/// A non-comparative sorting algorithm that sorts integers by processing individual digits.
/// It uses counting sort as a subroutine to sort based on each digit position.
fn radix_sort(arr: &mut [i32]) {
    if arr.is_empty() {
        return;
    }

    // Find the maximum number to know number of digits
    let max = *arr.iter().max().unwrap();
    
    // Do counting sort for every digit
    let mut exp = 1;
    while max / exp > 0 {
        counting_sort_for_radix(arr, exp);
        exp *= 10;
    }
}

/// Counting sort for each digit position
fn counting_sort_for_radix(arr: &mut [i32], exp: i32) {
    let n = arr.len();
    let mut output = vec![0; n];
    let mut count = [0; 10]; // count array for digits 0-9
    
    // Store count of occurrences in count[]
    for &x in arr.iter() {
        count[((x / exp) % 10) as usize] += 1;
    }
    
    // Change count[i] so that count[i] contains actual
    // position of this digit in output[]
    for i in 1..10 {
        count[i] += count[i - 1];
    }
    
    // Build the output array
    for &x in arr.iter().rev() {
        let digit = ((x / exp) % 10) as usize;
        output[count[digit] - 1] = x;
        count[digit] -= 1;
    }
    
    // Copy output array to arr[]
    arr.copy_from_slice(&output);
}

/// Reads numbers from file and returns as a vector.
fn read_numbers_from_file(filename: &str) -> io::Result<Vec<i32>> {
    let file = File::open(filename)?;
    let reader = BufReader::new(file);
    
    let numbers: Result<Vec<i32>, _> = reader
        .lines()
        .map(|line| line?.trim().parse::<i32>().map_err(|e| io::Error::new(io::ErrorKind::InvalidData, e)))
        .collect();
    
    numbers
}

/// Verifies if the array is sorted.
fn is_sorted<T: Ord>(arr: &[T]) -> bool {
    arr.windows(2).all(|window| window[0] <= window[1])
}

fn main() -> io::Result<()> {
    println!("Rust Radix Sort Performance Test");
    println!("================================");
    
    // Parse command line arguments
    let args: Vec<String> = env::args().collect();
    let filename = args.get(1).map_or("random_list.txt", |s| s.as_str());
    let result_filename = args.get(2).map_or("results_rust.txt", |s| s.as_str());
    
    println!("Using dataset: {}", filename);
    
    // Read data from file
    println!("Reading data from file...");
    let mut data = read_numbers_from_file(filename)?;
    println!("Data size: {} integers", data.len());
    
    // Create a copy for sorting (to preserve original)
    let mut data_copy = data.clone();
    
    // Measure sorting time
    println!("Starting Radix Sort...");
    let start = Instant::now();
    radix_sort(&mut data_copy);
    let duration = start.elapsed();
    
    let execution_time = duration.as_secs_f64();
    
    // Verify the array is sorted
    let sorted = is_sorted(&data_copy);
    
    // Results
    println!("Sorting completed: {}", if sorted { "SUCCESS" } else { "FAILED" });
    println!("Execution time: {:.6} seconds", execution_time);
    println!("Elements per second: {:.0}", data.len() as f64 / execution_time);
    
    // Save results to file
    let mut result_file = File::create(result_filename)?;
    writeln!(result_file, "Rust Radix Sort Results")?;
    writeln!(result_file, "Data size: {}", data.len())?;
    writeln!(result_file, "Execution time: {:.6} seconds", execution_time)?;
    writeln!(result_file, "Elements per second: {:.0}", data.len() as f64 / execution_time)?;
    writeln!(result_file, "Sorted correctly: {}", sorted)?;
    
    println!("Results saved to {}", result_filename);
    
    Ok(())
} 

File: compare-sort-algorithms/src/radix_sort.php
#!/usr/bin/env php
<?php

/**
 * Radix Sort Implementation in PHP
 * 
 * This script implements radix sort algorithm with proper error handling,
 * memory management, and execution time tracking.
 */

// Set memory limit to 1GB to handle large arrays
ini_set('memory_limit', '1G');

// Set maximum execution time to 300 seconds (5 minutes)
set_time_limit(300);

/**
 * Counting sort for a specific digit
 * 
 * @param array $arr The array to be sorted
 * @param int $exp The exponent (digit position)
 * @return array The sorted array for the current digit
 */
function countingSortForRadix(array $arr, int $exp): array {
    $n = count($arr);
    $output = array_fill(0, $n, 0);
    $count = array_fill(0, 10, 0);
    
    // Store count of occurrences in count[]
    for ($i = 0; $i < $n; $i++) {
        $index = (int)($arr[$i] / $exp) % 10;
        $count[$index]++;
    }
    
    // Change count[i] so that count[i] now contains actual
    // position of this digit in output[]
    for ($i = 1; $i < 10; $i++) {
        $count[$i] += $count[$i - 1];
    }
    
    // Build the output array
    for ($i = $n - 1; $i >= 0; $i--) {
        $index = (int)($arr[$i] / $exp) % 10;
        $output[$count[$index] - 1] = $arr[$i];
        $count[$index]--;
    }
    
    return $output;
}

/**
 * Radix Sort Implementation
 * 
 * @param array $arr The array to be sorted
 * @return array The sorted array
 * @throws Exception If input is invalid or memory limit is exceeded
 */
function radixSort(array $arr): array {
    $n = count($arr);
    if ($n === 0) {
        return $arr;
    }

    // Find the maximum number to know number of digits
    $max = max($arr);
    
    // Do counting sort for every digit
    $exp = 1;
    while ($max / $exp > 0) {
        $arr = countingSortForRadix($arr, $exp);
        $exp = (int)($exp * 10); // Ensure exp remains an integer
    }
    
    return $arr;
}

// Main execution block
if (php_sapi_name() === 'cli') {
    try {
        // Check command line arguments
        if ($argc < 2) {
            throw new Exception("Usage: php radix_sort.php <input_file> [output_file]");
        }

        // Read input file
        $input_file = $argv[1];
        if (!file_exists($input_file)) {
            throw new Exception("Input file not found: $input_file");
        }

        // Read numbers from file
        $numbers = array_map('intval', file($input_file, FILE_IGNORE_NEW_LINES | FILE_SKIP_EMPTY_LINES));
        
        // Sort the array
        $startTime = microtime(true);
        $sorted = radixSort($numbers);
        $endTime = microtime(true);
        
        // Calculate metrics
        $executionTime = ($endTime - $startTime); // in seconds
        $dataSize = count($numbers);
        $elementsPerSecond = $dataSize / $executionTime;
        
        // Verify if array is sorted correctly
        $isSorted = true;
        for ($i = 1; $i < count($sorted); $i++) {
            if ($sorted[$i] < $sorted[$i - 1]) {
                $isSorted = false;
                break;
            }
        }
        
        // Output metrics
        fprintf(STDERR, "PHP Radix Sort Results\n");
        fprintf(STDERR, "Data size: %d\n", $dataSize);
        fprintf(STDERR, "Execution time: %.6f seconds\n", $executionTime);
        fprintf(STDERR, "Elements per second: %.0f\n", $elementsPerSecond);
        fprintf(STDERR, "Sorted correctly: %s\n", $isSorted ? "true" : "false");
        fprintf(STDERR, "Memory usage: %.2f MB\n", memory_get_peak_usage(true) / 1024 / 1024);
        
    } catch (Exception $e) {
        fprintf(STDERR, "Error: %s\n", $e->getMessage());
        exit(1);
    }
} 